File: Hard/2026-04-25-3464-Maximize-the-Distance-Between-Points-on-a-Square/solution.java
import java.util.Arrays;

public class Solution {
    public int maxDistance(int side, int[][] points, int k) {
        int n = points.length;
        Arrays.sort(points, (a, b) -> Integer.compare(a[0] + a[1], b[0] + b[1]));
        
        int left = 0, right = 2 * side, answer = 0;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (canPlace(points, k, mid)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return answer;
    }

    private boolean canPlace(int[][] points, int k, int minDist) {
        int count = 1;
        int[] lastPoint = points[0];

        for (int i = 1; i < points.length; i++) {
            if (manhattanDistance(lastPoint, points[i]) >= minDist) {
                count++;
                lastPoint = points[i];
                if (count == k) {
                    return true;
                }
            }
        }
        return false;
    }

    private int manhattanDistance(int[] p1, int[] p2) {
        return Math.abs(p1[0] - p2[0]) + Math.abs(p1[1] - p2[1]);
    }
}

File: Hard/2026-04-25-3464-Maximize-the-Distance-Between-Points-on-a-Square/solution.py
from itertools import combinations

class Solution:
    def maxDistance(self, side: int, points: List[List[int]], k: int) -> int:
        def manhattan_distance(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
        
        def can_place(points, k, min_dist):
            count = 1
            last_point = points[0]
            for i in range(1, len(points)):
                if manhattan_distance(last_point, points[i]) >= min_dist:
                    count += 1
                    last_point = points[i]
                    if count == k:
                        return True
            return False
        
        points.sort()
        left, right = 0, 2 * side
        
        while left < right:
            mid = (left + right + 1) // 2
            if can_place(points, k, mid):
                left = mid
            else:
                right = mid - 1
        
        return left